moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,1254 @@
+}
+
+// A separate thread for the heavy jogging calculations.
+void* JogROSInterface::jogCalcThread(void*)
+{
+  jog_arm::JogCalcs ja(ros_parameters_, shared_variables_, model_loader_ptr_);
+  return nullptr;
+}
+
+// A separate thread for collision checking.
+void* JogROSInterface::collisionCheckThread(void*)
+{
+  jog_arm::collisionCheckThread cc(ros_parameters_, shared_variables_, model_loader_ptr_);
+  return nullptr;
+}
+
+// Constructor for the class that handles collision checking
+collisionCheckThread::collisionCheckThread(
+    const JogArmParameters& parameters, jog_arm_shared& shared_variables,
+    const std::unique_ptr<robot_model_loader::RobotModelLoader>& model_loader_ptr)
+{
+  // If user specified true in yaml file
+  if (parameters.collision_check)
+  {
+    // MoveIt Setup
+    // Wait for model_loader_ptr to be non-null.
+    while (ros::ok() && !model_loader_ptr)
+    {
+      ROS_WARN_THROTTLE_NAMED(5, NODE_NAME, "Waiting for a non-null robot_model_loader pointer");
+      ros::Duration(0.1).sleep();
+    }
+    const robot_model::RobotModelPtr& kinematic_model = model_loader_ptr->getModel();
+    planning_scene::PlanningScene planning_scene(kinematic_model);
+    collision_detection::CollisionRequest collision_request;
+    collision_request.group_name = parameters.move_group_name;
+    collision_request.distance = true;
+    collision_detection::CollisionResult collision_result;
+    robot_state::RobotState& current_state = planning_scene.getCurrentStateNonConst();
+    moveit::planning_interface::PlanningSceneInterface planning_scene_interface;
+
+    // Wait for initial messages
+    ROS_INFO_NAMED(NODE_NAME, "Waiting for first joint msg.");
+    ros::topic::waitForMessage<sensor_msgs::JointState>(parameters.joint_topic);
+    ROS_INFO_NAMED(NODE_NAME, "Received first joint msg.");
+
+    ROS_INFO_NAMED(NODE_NAME, "Waiting for first command msg.");
+    ros::topic::waitForMessage<geometry_msgs::TwistStamped>(parameters.cartesian_command_in_topic);
+    ROS_INFO_NAMED(NODE_NAME, "Received first command msg.");
+
+    // A very low cutoff frequency
+    jog_arm::LowPassFilter velocity_scale_filter(20);
+    // Assume no scaling, initially
+    velocity_scale_filter.reset(1);
+    ros::Rate collision_rate(parameters.collision_check_rate);
+
+    /////////////////////////////////////////////////
+    // Spin while checking collisions
+    /////////////////////////////////////////////////
+    while (ros::ok())
+    {
+      pthread_mutex_lock(&shared_variables.joints_mutex);
+      sensor_msgs::JointState jts = shared_variables.joints;
+      pthread_mutex_unlock(&shared_variables.joints_mutex);
+
+      for (std::size_t i = 0; i < jts.position.size(); ++i)
+        current_state.setJointPositions(jts.name[i], &jts.position[i]);
+
+      // process collision objects in scene
+      std::map<std::string, moveit_msgs::CollisionObject> c_objects_map = planning_scene_interface.getObjects();
```

Comment:
i think the planning scene monitor would handle for you all the processing of newly received collision objects. why is this being done so manual here?